pkg/certificate/providers/mrc.go

```diff
@@ -0,0 +1,118 @@
+package providers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/openservicemesh/osm/pkg/apis/config/v1alpha2"
+	"github.com/openservicemesh/osm/pkg/certificate"
+	"github.com/openservicemesh/osm/pkg/k8s/informers"
+)
+
+var (
+	errNilInformerCollection = errors.New("informer collection is nil")
+)
+
+var mrcKeyFunc = func(mrc v1alpha2.MeshRootCertificate) string {
+	return fmt.Sprintf("%s/%s", mrc.GetNamespace(), mrc.GetName())
+}
+
+// MRCComposer is a composer object that allows consumers
+// to observe MRCs (via List() and Watch()) as well as generate
+// `certificate.Provider`s from those MRCs
+type MRCComposer struct {
+	informerCollection *informers.InformerCollection
+	MRCProviderGenerator
+}
+
+// List returns the MRCs stored in the informerCollection's store
+func (m *MRCComposer) List() ([]*v1alpha2.MeshRootCertificate, error) {
+	if m.informerCollection == nil {
```

Comment:
this is a non-exported field, so we should use constructors such that this can never be nil, then we don't need the extra check

Reply:
Technically, the informerCollection still could be nil if its parameter higher in the stack is set to nil (e.g. during tests). These checks were honestly a way to get around nil pointer exceptions in a couple of tests, but I don't _think_ it's needed anymore because we've got a check upstream. I'll remove it